src/librbd/migration/NBDStream.cc

```diff
@@ -0,0 +1,235 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "librbd/migration/NBDStream.h"
+#include "common/dout.h"
+#include "common/errno.h"
+#include "librbd/AsioEngine.h"
+#include "librbd/ImageCtx.h"
+#include <boost/asio/buffer.hpp>
+
+#undef FMT_HEADER_ONLY
+#define FMT_HEADER_ONLY 1
+#include <fmt/chrono.h>
+#include <fmt/format.h>
+
+#include <time.h>
+
+namespace librbd {
+namespace migration {
+
+namespace {
+
+const std::string URL_KEY  {"url"};
+const std::string PORT_KEY {"port"};
+
+} // anonymous namespace
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::migration::NBDStream: " << this \
+                           << " " << __func__ << ": "
+
+template <typename I>
+struct NBDStream<I>::ReadRequest {
+  NBDStream*  nbd_stream;
+  io::Extents byte_extents;
+  bufferlist* data;
+  Context* on_finish;
+
+  ReadRequest(NBDStream* nbd_stream, io::Extents&& byte_extents,
+              bufferlist* data, Context* on_finish)
+    : nbd_stream(nbd_stream), byte_extents(std::move(byte_extents)),
+      data(data), on_finish(on_finish) {
+    auto cct = nbd_stream->m_cct;
+    ldout(cct, 20) << dendl;
+  }
+
+  void send() {
+    data->clear();
+    read();
+  }
+
+  void read() {
+    auto cct = nbd_stream->m_cct;  
+    struct nbd_handle *nbd = nbd_stream->nbd;
+
+    ldout(cct, 20) << "byte_extents=" << byte_extents << dendl;
+
+    for (auto [byte_offset, byte_length] : byte_extents) {
+      ldout(cct, 20) << "byte_offset=" << byte_offset << dendl;
+      ldout(cct, 20) << "byte_length=" << byte_length << dendl;
+      auto ptr = buffer::ptr_node::create(buffer::create_small_page_aligned(
+        byte_length));
+      auto buffer = boost::asio::mutable_buffer(ptr->c_str(), byte_length);
+      data->push_back(std::move(ptr));
+      auto rc = nbd_aio_pread(nbd, boost::asio::buffer_cast<void *>(buffer),
+        byte_length, byte_offset, NBD_NULL_COMPLETION, 0);
+      if(rc == -1) {
+        lderr(cct) << "nbd_aio_pread: " << nbd_get_error() << 
+                        " (errno=" << nbd_get_errno() << ")" <<  dendl;
+        on_finish->complete(-EINVAL);
+        return;
+      }
+    }
+
+    while (nbd_aio_in_flight(nbd) > 0) {
+      if (nbd_poll (nbd, -1) == -1) {
```

Comment:
It still does not make it async, as you are blocking here for the request to complete. What I meant is to register a callback in `nbd_aio_pread` (instead of `NBD_NULL_COMPLETION`), which would check the return value and call `on_finish->complete(r)` (`on_finish` could be stored in  `user_data` in `nbd_completion_callback`).

Reply:
The libnbd library does not spawn its own threads, so the code spawns a new thread in the read() function.  The read function then immediately returns without waiting for the actual processing of the read in ReadRequest.  This is similar to how the File stream handles aio.  But in addition, I switched to using the nbd_completion_callback.  Is that resonable?

Reply:
I am not quite sure I understand what you mean by "the code spawns a new thread in the read() function".

Anyway, My idea is that if you set a read callback you don't need call `nbd_poll` to wait until the read is complete and may just return immediately. And when the callback is callet it will complete the `on_finish` context. At least this is how I suppose the nbd_aio API could be used here (I have not tried it myself).

Reply:
Sorry if I wasn't clear.  The call to boost::asio::post(m_strand, [ctx]() { ctx->send_read(); }); in NBDStream::read() immediately returns control, so we are not blocking.  But regardless, I still believe we need the nbd_poll. Libnbd uses an event loop (see https://manpages.ubuntu.com/manpages/focal/man3/libnbd.3.html) and we either call nbd_poll or implement our our own version of checking for events.  BTW, I updated the code in question and pushed an update.  If you'll take a look you'll see I switched to using the callback, but I still needed to keep the nbd_poll call.  Without it, we never get into the callback.

Reply:
That sounds weird to me.  I don't have experience with libnbd but but your current aio version does not make much sense to me. I don't see why to use aio read if you are still blocking on poll. Then your initial "sync" version would be easier.

I am going to try your code this weekend and see what is going there.